File: backend/app/schemas/task.py
from typing import Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel, Field

TaskStatus = Literal[
    "created",
    "queued",
    "running",
    "stopping",
    "stopped",
    "success",
    "failed",
    "retrying",
    "partial_success",
]


class SystemPaths(BaseModel):
    gs_repo: str = "third_party/gaussian-splatting"
    raw_data: str = "datasets/raw"
    processed_data: str = "datasets/processed"
    outputs: str = "outputs"
    logs: str = "logs"
    videos_data: str = "datasets/videos"


class PipelineFlags(BaseModel):
    input_mode: Literal["images", "video"] = "images"
    run_preflight: bool = True
    run_video_extract: bool = False
    run_colmap: bool = True
    run_convert: bool = True
    run_train: bool = True
    run_render: bool = True
    run_metrics: bool = True
    launch_viewer: bool = False


class TrainProfile(BaseModel):
    active_profile: str = "low_vram"
    eval: bool = True
    iterations: int = 30000
    save_iterations: List[int] = Field(default_factory=lambda: [7000, 30000])
    test_iterations: List[int] = Field(default_factory=lambda: [-1])
    checkpoint_iterations: List[int] = Field(
        default_factory=lambda: [2000, 15000, 30000]
    )
    start_checkpoint: Optional[str] = ""
    resume_from_latest: bool = False
    quiet: bool = False
    extra_args: Dict[str, Any] = Field(
        default_factory=lambda: {
            "data_device": "cpu",
            "resolution": 4,
            "densify_grad_threshold": 0.001,
            "densification_interval": 200,
            "densify_until_iter": 3000,
        }
    )

class SceneConfig(BaseModel):
    scene_name: str
    raw_image_path: str = ""
    processed_scene_path: str = ""
    source_path: str = ""
    model_output: str = ""
    video_path: str = ""
    colmap_executable: str = "colmap"
    magick_executable: str = ""
    ffmpeg_executable: str = "ffmpeg"
    viewer_root: str = "third_party/viewer/bin"


class TaskCreateRequest(BaseModel):
    scene: SceneConfig
    system_paths: SystemPaths = Field(default_factory=SystemPaths)
    pipeline: PipelineFlags = Field(default_factory=PipelineFlags)
    train: TrainProfile = Field(default_factory=TrainProfile)


class StageRecord(BaseModel):
    stage_key: str
    stage_label: str
    order: int = 0
    status: Literal["pending", "running", "success", "failed", "stopped"] = "pending"
    started_at: Optional[str] = None
    finished_at: Optional[str] = None
    duration_seconds: Optional[float] = None
    error_type: Optional[str] = None
    error_message: Optional[str] = None


class TaskResponse(BaseModel):
    task_id: str
    scene_name: str
    status: Union[TaskStatus, str]
    current_stage: str
    message: str
    result: Dict[str, Any] = Field(default_factory=dict)
    error: Optional[str] = None

    created_at: Optional[str] = None
    started_at: Optional[str] = None
    finished_at: Optional[str] = None

    stop_requested: bool = False
    retry_count: int = 0

    stage_history: List[StageRecord] = Field(default_factory=list)
    metrics_summary: Dict[str, Any] = Field(default_factory=dict)
    result_files: Dict[str, Any] = Field(default_factory=dict)


class TaskActionResponse(BaseModel):
    ok: bool = True
    task_id: str
    action: Literal["stop", "retry", "delete"]
    status: Union[TaskStatus, str]
    message: str

class TaskLogResponse(BaseModel):
    task_id: str
    lines: List[str] = Field(default_factory=list)
    count: int = 0

File: frontend/src/types/task.ts
export type 任务状态 =
  | 'created'
  | 'queued'
  | 'running'
  | 'stopping'
  | 'stopped'
  | 'success'
  | 'failed'
  | 'retrying'
  | 'partial_success'

export type 阶段状态 =
  | 'pending'
  | 'running'
  | 'success'
  | 'failed'
  | 'stopped'

export type 数据增强预设 = 'off' | 'safe' | 'low_light' | 'detail' | 'custom'

export interface 阶段记录 {
  stage_key: string
  stage_label: string
  order: number
  status: 阶段状态 | string
  started_at: string | null
  finished_at: string | null
  duration_seconds: number | null
  error_type: string | null
  error_message: string | null
}

export interface 系统路径配置 {
  gs_repo: string
  raw_data: string
  processed_data: string
  outputs: string
  logs: string
  videos_data: string
}

export interface 数据增强配置 {
  enabled: boolean
  preset: 数据增强预设
  output_subdir: string
  overwrite: boolean
  keep_original_if_failed: boolean
  jpeg_quality: number

  gray_world: boolean
  clahe: boolean
  clahe_clip_limit: number
  clahe_tile_grid_size: [number, number]

  auto_gamma: boolean
  gamma_target_mean: number

  denoise: boolean
  denoise_h: number

  sharpen: boolean
  sharpen_amount: number

  /** 0 表示不限制长边；大于 0 时按最长边等比例缩小，避免图片过大导致 COLMAP / 训练显存压力过高。 */
  max_long_edge: number
}

export interface 创建任务请求 {
  scene: {
    scene_name: string
    raw_image_path: string
    processed_scene_path: string
    source_path: string
    model_output: string
    video_path: string
    colmap_executable: string
    magick_executable: string
    ffmpeg_executable: string
    viewer_root: string
    colmap_use_gpu: boolean
    video_target_fps: number
  }
  system_paths: 系统路径配置
  pipeline: {
    input_mode: 'images' | 'video'
    run_preflight: boolean
    run_video_extract: boolean
    run_augmentation: boolean
    run_colmap: boolean
    run_convert: boolean
    run_train: boolean
    run_render: boolean
    run_metrics: boolean
    launch_viewer: boolean
  }
  augmentation: 数据增强配置
  train: {
    active_profile: string
    eval: boolean
    iterations: number
    save_iterations: number[]
    test_iterations: number[]
    checkpoint_iterations: number[]
    start_checkpoint: string
    resume_from_latest: boolean
    quiet: boolean
    extra_args: {
      data_device: string
      resolution: number
      densify_grad_threshold: number
      densification_interval: number
      densify_until_iter: number
      [key: string]: string | number | boolean | null
    }
  }
}

export interface 任务响应 {
  task_id: string
  scene_name: string
  status: 任务状态 | string
  current_stage: string
  message: string
  result: Record<string, unknown>
  error: string | null

  created_at: string | null
  started_at: string | null
  finished_at: string | null

  stop_requested: boolean
  force_stop_requested: boolean
  retry_count: number

  stage_history: 阶段记录[]
  metrics_summary: Record<string, unknown>
  result_files: Record<string, unknown>
}

export interface 任务动作响应 {
  ok: boolean
  task_id: string
  action: 'stop' | 'force_stop' | 'retry' | 'delete'
  status: 任务状态 | string
  message: string
}

export interface 任务列表响应 {
  items: 任务响应[]
}

export interface 结果响应 {
  task_id: string
  scene_name: string
  status: 任务状态 | string
  current_stage: string
  message: string
  error: string | null

  created_at: string | null
  started_at: string | null
  finished_at: string | null

  stop_requested: boolean
  force_stop_requested: boolean
  retry_count: number

  stage_history: 阶段记录[]
  metrics_summary: Record<string, unknown>
  result_files: Record<string, unknown>
  result: Record<string, unknown>
}

export interface 任务日志响应 {
  task_id: string
  lines: string[]
  count: number
}
